server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/Repository.java

```diff
@@ -424,4 +434,62 @@ default Author author() {
     default <T> CompletableFuture<Entry<T>> watch(Revision lastKnownRevision, Query<T> query) {
         return RepositoryUtil.watch(this, lastKnownRevision, query);
     }
+
+    /**
+     * Merges the JSON files sequentially as specified in the {@link MergeQuery}.
+     */
+    default <T> CompletableFuture<MergedEntry<T>> mergeFiles(Revision revision, MergeQuery<T> query) {
+        requireNonNull(revision, "revision");
+        requireNonNull(query, "query");
+
+        final List<MergeSource> paths = query.mergeSources();
+        // Only JSON files can currently be merged.
+        paths.forEach(path -> validateJsonFilePath(path.path(), "path"));
+
+        final Revision normalizedRevision = normalizeNow(revision);
+        final List<CompletableFuture<Entry<?>>> entryFutures = new ArrayList<>(Iterables.size(paths));
+        paths.forEach(path -> {
+            if (!path.isOptional()) {
+                entryFutures.add(get(normalizedRevision, path.path()));
+            } else {
+                entryFutures.add(getOrNull(normalizedRevision, path.path()));
+            }
+        });
+
+        final CompletableFuture<MergedEntry<JsonNode>> future = new CompletableFuture<>();
+        CompletableFutures.allAsList(entryFutures).handle((entries, cause) -> {
+            if (cause != null) {
+                future.completeExceptionally(cause);
+            }
+
+            final Builder<JsonNode> builder = ImmutableList.builder();
+            for (Entry<?> entry : entries) {
+                if (entry == null) {
+                    continue;
+                }
+                try {
+                    builder.add(entry.contentAsJson());
+                } catch (JsonParseException e) {
+                    future.completeExceptionally(e);
+                    return null;
+                }
+            }
+
+            JsonNode result;
+            try {
+                result = Jackson.mergeTree(builder.build());
```

Comment:
I wonder if we should pass an empty list if all merge sources are non-existent. Should we raise a `QueryExecutionException` rather than `IllegalArgumentException`?